hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/infra/StorageAccessTracker.java

```diff
@@ -0,0 +1,84 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.contract.impl.infra;
+
+import com.hedera.hapi.node.state.contract.SlotKey;
+import com.hedera.node.app.service.contract.impl.state.StorageAccess;
+import com.hedera.node.app.service.contract.impl.state.StorageAccesses;
+import com.hedera.services.stream.proto.SidecarType;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.function.Function;
+import org.apache.tuweni.units.bigints.UInt256;
+
+/**
+ * Tracks storage accesses for a particular transaction.
+ */
+public class StorageAccessTracker {
+    private static final Function<Long, Map<UInt256, StorageAccess>> MAP_FACTORY = ignored -> new TreeMap<>();
+    private final Map<Long, Map<UInt256, StorageAccess>> accessesByContract = new TreeMap<>();
+
+    /**
+     * The first time this method is called for a particular {@link SlotKey}, tracks its
+     * value for future reporting in a {@link SidecarType#CONTRACT_STATE_CHANGE} sidecar.
+     *
+     * @param number the number of the contract whose storage is being read
+     * @param key the key of the slot read
+     * @param value the value read
+     */
+    public void trackIfFirstRead(final long number, @NonNull final UInt256 key, @NonNull final UInt256 value) {
+        accessesByContract.computeIfAbsent(number, MAP_FACTORY).putIfAbsent(key, StorageAccess.newRead(key, value));
+    }
+
+    /**
+     * Returns the list of all storage reads (i.e. the tracked {@code SLOAD}'s).
+     * This is a convenience methods equivalent to passing an empty list to
+     * {@link #getReadsMergedWith(List)}
+     *
+     * @return the list of all storage reads
+     */
+    public List<StorageAccesses> getJustReads() {
+        return getReadsMergedWith(List.of());
+    }
+
+    /**
+     * Given all the storage writes from a transaction, returns the merged list of {@link StorageAccesses} that
+     * includes both the given list of writes, and all tracked first reads.
+     *
+     * <p>If a key was both read and overwritten, the read-only access will not appear,
+     * since a write {@link com.hedera.node.app.service.contract.impl.state.StorageAccess}
+     * already includes the overwritten value.
+     *
+     * @param writes all the storage
+     * @return the merged list of all storage accesses
+     */
+    public List<StorageAccesses> getReadsMergedWith(@NonNull final List<StorageAccesses> writes) {
```

Comment:
Nicely compact/terse.  Needs a bit of concentration to read it but it all becomes clear ...